pkg/oauth/oauth.go

```diff
@@ -52,10 +56,61 @@ type TokenStrategy struct {
 // TokenStrategySettings represents the settings for the token strategy
 type TokenStrategySettings map[string]string
 
+// GetJWTSecret gets the JWT secret config
+func (t TokenStrategySettings) GetJWTSecret() (string, error) {
+	value, ok := t["secret"]
+	if !ok || value == "" {
+		return "", ErrJWTSecretMissing
+	}
+
+	return value, nil
+}
+
 type meta struct {
 	Domains        []string `mapstructure:"domains" bson:"domains" json:"domains"`
 	Methods        []string `mapstructure:"methods" bson:"methods" json:"methods"`
 	RequestHeaders []string `mapstructure:"request_headers" bson:"request_headers" json:"request_headers"`
 	ExposedHeaders []string `mapstructure:"exposed_headers" bson:"exposed_headers" json:"exposed_headers"`
 	Enabled        bool     `bson:"enabled" json:"enabled"`
 }
+
+// AccessRule represents a rule that will be applied to a JWT that could be revoked
+type AccessRule struct {
+	Predicate string `bson:"predicate" json:"predicate"`
+	Action    string `bson:"action" json:"action"`
+	parsed    bool
+}
+
+// IsAllowed checks if the rule is allowed to
+func (r *AccessRule) IsAllowed(claims map[string]interface{}) (bool, error) {
+	var err error
+
+	if !r.parsed {
+		matched, err := r.parse(claims)
+		if err != nil {
+			return false, err
+		}
+
+		if !matched {
+			return true, nil
+		}
+	}
+
+	return r.Action == "allow", err
+}
+
+func (r *AccessRule) parse(claims map[string]interface{}) (bool, error) {
+	expression, err := govaluate.NewEvaluableExpression(r.Predicate)
+	if err != nil {
+		return false, errors.New("Could not create an expression with this predicate")
+	}
+
+	result, err := expression.Evaluate(claims)
+	if err != nil {
+		return false, errors.New("Cannot evaluate the expression")
+	}
+
+	r.parsed = true
```

Comment:
As this will be called on every request we should we should make it goroutine safe:
```go
r.mu.Lock()
r.parsed = true
r.mu.UnLock()
....
```